### PR TITLE
LLM: Enable BigDL IPEX optimization for int4

### PR DESCRIPTION
**Pull request by xiangyuT**
## Description

Enable BigDL IPEX optimization for int4

### CODE REVIEW

**Review by glorysdj — APPROVED** [review]
LGTM

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 1]
```diff
@@ -671,12 +670,19 @@ def replace_func(m, target_m, func_name, new_func):
         replace_func(sub_m, target_m, func_name, new_func)
 
 
-def _optimize_ipex(model):
+def get_enable_ipex():
+    _enable_ipex = os.getenv("BIGDL_OPT_IPEX")
+    _enable_ipex = (_enable_ipex is not None) and (_enable_ipex.lower() == "true")
+    return _enable_ipex
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/convert.py:676]: only bf16 or int4 is supported, yes?

File: python/llm/src/bigdl/llm/transformers/convert_ipex.py [hunk 2]
```diff
@@ -91,18 +95,47 @@ def _ipex_optimize_attention(model):
...
+            "PER_BATCH": ipex.quantization.WoqActQuantMode.PER_BATCH,
+            "PER_BATCH_IC_BLOCK": ipex.quantization.WoqActQuantMode.PER_BATCH_IC_BLOCK,
+        }
+        qconfig = ipex.quantization.get_weight_only_quant_qconfig_mapping(
+            weight_dtype=torch.quint4x2,  # INT4
+            lowp_mode=ipex.quantization.WoqLowpMode.INT8,
+            act_quant_mode=act_quant_mode_dict["PER_IC_BLOCK"],
+            group_size=-1,
+        )
+        model = ipex_quantization_flow(model, torch.bfloat16, None, qconfig, None)
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/convert_ipex.py:129]: Only llama is verified, yes?

File: python/llm/src/bigdl/llm/transformers/convert.py (hunk 1 above)
Comment (xiangyuT) [reply python/llm/src/bigdl/llm/transformers/convert.py:676]: > only bf16 or int4 is supported, yes?

yes

File: python/llm/src/bigdl/llm/transformers/model.py [hunk 3]
```diff
@@ -408,7 +408,14 @@ def load_convert(cls, q_k, optimize_model, *args, **kwargs):
...
             try:
-                model = cls.HF_Model.from_pretrained(*args, **kwargs)
+                from bigdl.llm.transformers.convert import get_enable_ipex
+                _enable_ipex = get_enable_ipex()
+                if _enable_ipex:
+                    import intel_extension_for_pytorch as ipex
+                    with ipex.OnDevice(dtype=torch.float, device="meta"):
+                        model = cls.HF_Model.from_pretrained(*args, **kwargs)
+                else:
+                    model = cls.HF_Model.from_pretrained(*args, **kwargs)
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/model.py]: Can we have something like
```
with bigdl.llm.transformers.convert.get_context(...)
    cls.HF_Model.from_pretrained(...)
```

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 4]
```diff
@@ -596,7 +603,28 @@ def speculative_generate(self,
...
+                            attention_mask=draft_attention_mask,
+                            position_ids=position_ids,
+                            past_key_values=draft_past_key_values,
+                        )
+                    else:
+                        invalidInputError(False, "Only support Llama yet.")
+
+                    draft_output = CausalLMOutputWithPast(
+                        logits=draft_output[0],
+                        past_key_values=draft_output[1],
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/speculative.py:624]: How can one run speculative decoding for, say, ChatGLM3, using IPEX OPT now?

File: python/llm/src/bigdl/llm/transformers/convert_ipex.py (hunk 2 above)
Comment (xiangyuT) [reply python/llm/src/bigdl/llm/transformers/convert_ipex.py:129]: Ziteng and I has verified llama2, vicuna, chatglm3, baichuan2, qwen, mistral, starcoder and ziya model.

File: python/llm/src/bigdl/llm/transformers/speculative.py (hunk 4 above)
Comment (xiangyuT) [reply python/llm/src/bigdl/llm/transformers/speculative.py:624]: Should be similar to llama, I'm verifying speculative decoding for some other models and will update soon.

Comment (xiangyuT) [reply python/llm/src/bigdl/llm/transformers/speculative.py]: llama2, vicuna, chatglm3, baichuan2, qwen, mistral models have been verified on SPR.

**Pull request merged**